### PR TITLE
Empty accounts.json processing

### PR DESCRIPTION
**Pull request by velzevur**
Addressing some corner cases:

* Empty `accounts.json`

* `accounts.json` containing an empty js object: JSX has some curious behaviour with it ( `[{}] = jsx:decode(<<"{}">>).`) and it was causing crashes

Added EUnit tests to cover the functionality of the module

### CODE REVIEW

File: apps/aecore/src/aec_genesis_block_settings.erl [hunk 1]
```diff
@@ -1,26 +1,46 @@
...
-    case file:read_file(PresetAccountsFile) of
-        {error, _Err} ->
+    case ?MODULE:read_presets() of
+        {error, _Err, PresetAccountsFile} ->
             % no setup, no preset accounts
             erlang:error({genesis_accounts_file_missing, PresetAccountsFile});
         {ok, JSONData} ->
+            DecodedData =
+                case JSONData of
+                    <<"{}">> ->
```
Comment (lucafavatella) [inline apps/aecore/src/aec_genesis_block_settings.erl]: This looks fragile e.g. for whitespaces/newlines before/in-middle/after the `{}`.

File: apps/aecore/src/aec_genesis_block_settings.erl [hunk 2]
```diff
@@ -1,26 +1,44 @@
...
             % ensure deterministic ordering of accounts
             lists:keysort(1, Accounts)
     end.
+
+-spec read_presets() -> {ok, binary()}| {error, atom(), string()}.
+read_presets() ->
+    PresetAccountsFile = filename:join([dir(), "accounts.json"]),
+    case file:read_file(PresetAccountsFile) of
+        {ok, _} = OK -> OK;
+        {error, Err} -> {error, Err, PresetAccountsFile}
```
Comment (ThomasArts) [inline apps/aecore/src/aec_genesis_block_settings.erl]: Since dialyzer would not catch it, it might be better to stay to tuple notation {error, {Err, PresetAccountsFile}}

File: apps/aecore/src/aec_genesis_block_settings.erl [hunk 3]
```diff
@@ -1,26 +1,44 @@
...
 
 dir() ->
     filename:join(aeu_env:data_dir(aecore), ".genesis").
 
+-spec preset_accounts() -> list().
 preset_accounts() ->
-    PresetAccountsFile = filename:join([dir(), "accounts.json"]),
-    case file:read_file(PresetAccountsFile) of
-        {error, _Err} ->
+    case ?MODULE:read_presets() of
```
Comment (ThomasArts) [inline apps/aecore/src/aec_genesis_block_settings.erl]: Why ?MODULE here... it's the same module anyway?

File: apps/aecore/test/aec_genesis_block_settings_tests.erl [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+     end,
+     fun(ok) ->
+         meck:unload(?TEST_MODULE)
+     end,
+     [ {"Preset accounts parsing: broken file",
+        fun() ->
+            %% empty file
+            expect_accounts(<<"">>),
+            ?assertError(invalid_accounts_json, ?TEST_MODULE:preset_accounts()),
+            %% broekn json
```
Comment (ThomasArts) [inline apps/aecore/test/aec_genesis_block_settings_tests.erl]: broken

File: apps/aecore/src/aec_genesis_block_settings.erl (hunk 3 above)
Comment (velzevur) [reply apps/aecore/src/aec_genesis_block_settings.erl]: I know this is ugly but I could not find a way around it.
This module is being tested and in order to use `meck` for the `read_presets/1` - a fully qualified function call must be used. Other option would be meck-ing filename module...
I could add a comment for that if you find it useful

Comment (ThomasArts) [reply apps/aecore/src/aec_genesis_block_settings.erl]: I guess this is a misunderstanding of meck. You don't need to add the module in the call to make meck pick it up. Ask Adam for details.

Comment (eproxus) [reply apps/aecore/src/aec_genesis_block_settings.erl]: Just adding this here for posterity: Meck cannot mock local function calls in a module. The alternative is to use a fully qualified call (`Mod:Func(...)`) or even better, not to test using mock functions inside modules but at module borders only.

**Comment by velzevur** [comment]
Opening a new PR

**Comment by velzevur** [comment]
New PR #910
